cg/meta/upload/gisaid/models.py

```diff
@@ -58,22 +64,36 @@ class GisaidSample(BaseModel):
     covv_subm_lab_addr: Optional[str] = "171 76 Stockholm, Sweden"
     covv_authors: Optional[str] = " ,".join(AUTHORS)
 
-    @field_validator("covv_location")
+    @model_validator(mode="before")
     @classmethod
-    def parse_location(cls, _, info: FieldValidationInfo):
-        region: str = info.data.get("region")
-        return f"Europe/Sweden/{region}"
+    def set_generated_fields(cls, data: Any):
+        """Constructs the fields that are generated from other fields."""
+        if isinstance(data, dict):
+            data.setdefault("covv_location", _generate_covv_location(data.get("region")))
+            data["covv_subm_sample_id"] = _generate_covv_subm_sample_id(
+                subm_sample_id=data.get(
+                    "covv_subm_sample_id",
+                ),
+                region_code=data.get("region_code"),
+            )
+            data.setdefault(
+                "covv_virus_name",
+                _generate_covv_virus_name(
+                    covv_subm_sample_id=data.get("covv_subm_sample_id"),
+                    covv_collection_date=data.get("covv_collection_date"),
+                ),
+            )
+        return data
 
-    @field_validator("covv_subm_sample_id")
-    @classmethod
-    def parse_subm_sample_id(cls, v: str, info: FieldValidationInfo):
-        region_code = info.data.get("region_code")
-        return f"{region_code}_SE100_{v}"
 
-    @field_validator("covv_virus_name")
-    @classmethod
-    def parse_virus_name(cls, _, info: FieldValidationInfo):
-        sample_name = info.data.get("covv_subm_sample_id")
-        date = info.data.get("covv_collection_date")
-        datetime_date = datetime.strptime(date, "%Y-%m-%d")
-        return f"hCoV-19/Sweden/{sample_name}/{datetime_date.year}"
+def _generate_covv_location(region):
+    return f"Europe/Sweden/{region}"
+
+
+def _generate_covv_subm_sample_id(subm_sample_id, region_code):
```

Comment:
maybe add typed args and return type hints

Reply:
Fixed